Add importer for EUMETSAT SAF Files

cvelascof opened the pull request:
@juanpablosimarro did write an importer for SAF CRRI netcdf files, updated pysteps visualization scripts to display these datasets and create new examples using pySTEPS with Satellite data. 
https://github.com/juanpablosimarro/pysteps/tree/saf
 
This pull request includes only the CRRI importer and an additional test for the CRRI importer.
@juanpablosimarro and I will continue working to add more of his updates into pySTEPS.

This branch passed all tests in Travis for Linux but for some reason MacOS env seems broken at Travis.

cvelascof committed:
PEP8 changes

cvelascof committed:
Add netCDF SAF CRRI importer

cvelascof committed:
add crri-saf data source details

cvelascof committed:
Fix PEP8

cvelascof committed:
Add test to saf crri importer

cvelascof committed:
Fix incorrect call to module

cvelascof committed:
Fix dimensions

cvelascof committed:
Add test for crri_eu importer

cvelascof committed:
PEP8 corrections

cvelascof committed:
Fix dimensions product

aperezhortal committed:
Force installing gcc-9 in travis-ci

aperezhortal committed:
Run brew update-reset & brew update before installing gcc-9

This may fix the error during gcc installation in travis-ci for osx.

See:
- https://github.com/Homebrew/homebrew-core/issues/37269
- https://github.com/Homebrew/brew/issues/4755

aperezhortal committed:
Fix syntax error in previous commit

aperezhortal commented:
Hi @cvelascof, there was a problem with the brew installation. Running **brew update-reset** and **brew update** before installing gcc-9 solved the problem (commits 
0c3db1e  and 11490fc).

cvelascof commented:
Fantastic @aperezhortal. Thanks for fixing the macOs env.

juanpablosimarro commented:
Carlos and Andres, thank you.


El mar., 15 oct. 2019 a las 1:15, Carlos Velasco (<notifications@github.com>)
escribió:

> Fantastic @aperezhortal <https://github.com/aperezhortal>. Thanks for
> fixing the macOs env.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pySTEPS/pysteps/pull/127?email_source=notifications&email_token=AMTXTEXBB4O4CFQW5Y5NBZLQOT4Q7A5CNFSM4JAPOKA2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBG43QI#issuecomment-541969857>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMTXTEVAGZTTZMSVMM3ONZTQOT4Q7ANCNFSM4JAPOKAQ>
> .
>

aperezhortal commented on pysteps/io/importers.py:
```diff
@@ -135,6 +135,119 @@
+        metadata["threshold"] = np.nan
+
+    # cut a window, to be defined here and only here
+    if True:
```
This "if" condition is always executed. Is this there for a future keyword? If it is so, the "if" check may be removed and replaced by a TODO comment.

dnerini replied:
What is the motivation to include the domain clipping in the importer?
Personally I would prefer to keep this sort of data manipulation outside of the importers module. If this was motivated by memory limits, perhaps we could add a general option in the readers module, in order to apply some data processing operations (eg clipping, upscaling,..) while reading the data.

Besides, there is a utils method already implemented in pysteps to clip rasters based on geographical coordinates ([clip_domain()](https://pysteps.readthedocs.io/en/latest/generated/pysteps.utils.dimension.clip_domain.html#pysteps.utils.dimension.clip_domain)). Probably worth to have a look at it to see if it can already perform the operation you need.

dnerini replied:
solved by 99d6b65

aperezhortal committed:
Rename "R" variable as "precip"

juanpablosimarro commented:
Hello Andres,

Sorry for the delay.

Yes, one possibility is to define some geographical windows here through a
keyword (Europe, East Europe, West Europe, countries, etc). If you agree, I
can set a list of suitable windows here.

The keyword would select the geographical window or, alternatively let the
user to define its own. Nevertheless, those windows would be only valid for
those NETCDF SAF products I have used. The SAF software can produce the
same products with a different geographical extension (the METEOSAT
provides an image of almost one whole hemisphere).

Please tell me what you think.

King regards,

Juan




El mar., 15 oct. 2019 a las 21:04, Andres Perez Hortal (<
notifications@github.com>) escribió:

> *@aperezhortal* requested changes on this pull request.
> ------------------------------
>
> In pysteps/io/importers.py
> <https://github.com/pySTEPS/pysteps/pull/127#discussion_r335118888>:
>
> > +    R = _import_crri_eu_data(filename)
> +
> +    geodata = _import_crri_eu_geodata(filename)
> +    metadata = geodata
> +
> +    # TODO(import_crri_eu): Add missing georeferencing data.
> +
> +    metadata["transform"] = None
> +    metadata["zerovalue"] = np.nanmin(R)
> +    if np.any(np.isfinite(R)):
> +        metadata["threshold"] = np.nanmin(R[R > np.nanmin(R)])
> +    else:
> +        metadata["threshold"] = np.nan
> +
> +    # cut a window, to be defined here and only here
> +    if True:
>
> This "if" condition is always executed. Is this there for a future
> keyword? If it is so, the "if" check may be removed and replaced by a TODO
> comment.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pySTEPS/pysteps/pull/127?email_source=notifications&email_token=AMTXTES3WZXRCNREJR5GSTTQOYH2RA5CNFSM4JAPOKA2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCIBAHUI#pullrequestreview-302121937>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMTXTETZRKTB2SPLVHCHGV3QOYH2RANCNFSM4JAPOKAQ>
> .
>

juanpablosimarro commented:
Hello Daniele,

The projection and area of the original NETCDF file is that in the attached
image. It is very inconvenient for spectral methods, because pixels varies
a lot from one to another in shape and size, there are a lot of void space,
etc.

One possibility is the one you mention, that is, to use external tools to
cut and reproject this image to the user desired one. I tried to do this
and I did not find the way. I'm not expert in this kind of software (for
instance https://proj.org), could be you can help me. So, after trying some
options I found this simple, however not optimal: to cut the window in the
import sequence.

Another possibily you point out, is to use clip_domain(). I did know it
exists. Then if I'm not mistaken, the sequence would be: import the whole
NETCDF SAF files, and after that use clip_domain()
<https://pysteps.readthedocs.io/en/latest/generated/pysteps.utils.dimension.clip_domain.html#pysteps.utils.dimension.clip_domain>
before continuing with the calculations. I will try to use it.

Possibly the best option is this second one.

Best regards

Juan





El jue., 17 oct. 2019 a las 12:27, Daniele Nerini (<notifications@github.com>)
escribió:

> *@dnerini* commented on this pull request.
> ------------------------------
>
> In pysteps/io/importers.py
> <https://github.com/pySTEPS/pysteps/pull/127#discussion_r335928499>:
>
> > +    R = _import_crri_eu_data(filename)
> +
> +    geodata = _import_crri_eu_geodata(filename)
> +    metadata = geodata
> +
> +    # TODO(import_crri_eu): Add missing georeferencing data.
> +
> +    metadata["transform"] = None
> +    metadata["zerovalue"] = np.nanmin(R)
> +    if np.any(np.isfinite(R)):
> +        metadata["threshold"] = np.nanmin(R[R > np.nanmin(R)])
> +    else:
> +        metadata["threshold"] = np.nan
> +
> +    # cut a window, to be defined here and only here
> +    if True:
>
> What is the motivation to include the domain clipping in the importer?
> Personally I would prefer to keep this sort of data manipulation outside
> of the importers module. If this was motivated by memory limits, perhaps we
> could add a general option in the readers module, in order to apply some
> data processing operations (eg clipping, upscaling,..) while reading the
> data.
>
> Besides, there is a utils method already implemented in pysteps to clip
> rasters based on geographical coordinates (clip_domain()
> <https://pysteps.readthedocs.io/en/latest/generated/pysteps.utils.dimension.clip_domain.html#pysteps.utils.dimension.clip_domain>).
> Probably worth to have a look at it to see if it can already perform the
> operation you need.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pySTEPS/pysteps/pull/127?email_source=notifications&email_token=AMTXTERRZ7XZSUWN6LMTYH3QPA42PA5CNFSM4JAPOKA2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCII4PDI#discussion_r335928499>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMTXTERVQ5VCJ2XEI65Z2TTQPA42PANCNFSM4JAPOKAQ>
> .
>

dnerini commented on pysteps/io/interface.py:
```diff
@@ -42,51 +43,55 @@ def get_method(name, method_type):
+        +--------------+------------------------------------------------------+
+        | knmi_hdf5    |  HDF5 file format used by KNMI.                      |
+        +--------------+------------------------------------------------------+
+        | crri_eu      |  NetCDF SAF CRRI files                               |
```
Maybe we should try to keep this in alphabetical order. 
(can you also adjust the position of the knmi_hdf5 entry, please?)

dnerini commented on pysteps/io/importers.py:
```diff
@@ -135,6 +135,119 @@
     PYPROJ_IMPORTED = False
 
 
+def import_crri_eu(filename, **kwargs):
```
I would prefer to keep an alphabetical order for the importers in this module.

Also, don't forget to add your method's name in the list of available importers in the docstrings at the beginning of the module (lines 75-82).

dnerini reviewed (changes requested):
Excellent contribution, thanks!
I just included few minor comments and, as already discussed, I think we should move the windowing outside of the importer.

aperezhortal commented on pysteps/io/importers.py:
```diff
@@ -135,6 +135,119 @@
 
 
+def import_crri_eu(filename, **kwargs):
+    """Import a NetCDF radar rainfall product from the CRRI SAF.
```
Can you also add what SAF and CRRI stand form?

dnerini replied:
I can't find any SAF product named "CRRI", maybe you meant to write simply CRR (Convective Rain Rate)?

edit: uh maybe I got it: it's the CRR_INTENSITY product.. doh!

aperezhortal replied:
Added!

juanpablosimarro replied:
Your are right, inside the same file there are other products, for instance hourly accumulation

![Screenshot from 2019-11-08 12-03-51](https://user-images.githubusercontent.com/52918674/68472111-0ee02100-0220-11ea-9908-3a99a132ad0d.png)

dnerini commented:
> Hello Daniele, The projection and area of the original NETCDF file is that in the attached image. It is very inconvenient for spectral methods, because pixels varies a lot from one to another in shape and size, there are a lot of void space, etc. 

Hi @juanpablosimarro , thanks for your answer. Most of the methods in pysteps assume that pixels in your grid have constant size, hence the need to reproject your data.

> One possibility is the one you mention, that is, to use external tools to cut and reproject this image to the user desired one. I tried to do this and I did not find the way. I'm not expert in this kind of software (for instance https://proj.org), could be you can help me. 

Of course! This PR is probably not the right place to discuss this though. Maybe we can discuss this aspect further together on the pysteps slack.

> So, after trying some options I found this simple, however not optimal: to cut the window in the import sequence. Another possibily you point out, is to use clip_domain(). I did know it exists. Then if I'm not mistaken, the sequence would be: import the whole NETCDF SAF files, and after that use clip_domain() before continuing with the calculations. I will try to use it. Possibly the best option is this second one. Best regards 

I'm not sure to understand how clipping the domain is helping with the reprojection of the data. Perhaps because you focus on a subset of your domain where the pixels are reasonably regular?

Personally, I usually first read the raw data in and then adjust the domain at once with clip_domain(). However, this is not necessarily the most efficient way, particularly if you need to limit memory usage when working with large domains. 

In this case, I would suggest to include the option to apply clip_domain() to each individual raster while looping the timestamps in the read_timeseries() method. 

It is also true that it could be a good idea to directly extract a subset of the raster by leaving the option to provide lat lon bounds to the _import_crri_eu_data()  and then simply using indexes, i.e.

`data = np.array(ds_rainfall.variables["crr_intensity"][idx_lat, idx_lon])`.

This kind of contradicts my comment about removing the windowing from the importer, but I guess it would provide an efficient solution, especially with these large satellite rasters.

RubenImhoff commented on pysteps/io/importers.py:
```diff
@@ -135,6 +135,119 @@
+    ymax = py.max()
+    xpixelsize = abs(geotable[1])
+    ypixelsize = abs(geotable[5])
+    factor_scale = 1.0
```
Is this 'factor_scale' a changeble factor? If so, I would recommend to add it as e.g. '**kwargs' in the function (unless it's already somewhere defined and I've missed it..). If not, we won't need this variable at all, I assume.

dnerini replied:
removed in cd52933

RubenImhoff commented:
Thanks for the useful contribution by the way!

pulkkins committed:
Sort methods in alphabetical order

pulkkins committed:
Add import_bom_rf3 to the list of available importers

pulkkins committed:
Sort methods in alphabetical order

pulkkins commented:
I sorted the methods in io/interface.py and io/importers.py in alphabetical order.

aperezhortal committed:
Add CRRI and SAF meaning

dnerini committed:
Improve docstrings

dnerini committed:
Fix extent definition

- x1, x2, y1, y2 refer to the coordinates of the lower-left and upper-right corners of the domain, and not to the coordinates of actual pixels
- in the previous version, these coordinates were off by half pixel length

dnerini committed:
Remove factor_scale variable

dnerini committed:
Retrieve institute from file

dnerini committed:
Change name 'crri_eu' to 'saf_crri'

dnerini committed:
Add option to specify spatial extent of data

dnerini committed:
Fix wrong indent

dnerini committed:
Fix wrong import

dnerini committed:
Add missing arguments

dnerini committed:
Refactor test

dnerini commented:
After a (long expected) closer look, I realized that the SAF CRRI data are actually already projected, my apologies to @juanpablosimarro. This means that the precipitation data are on a regular grid and, in principle, ready to be used by pysteps!

I added few changes to the pull request, mainly added the option to specify the spatial extent when reading the data, instead of hard-coding it in the importer (99d6b65).

I also fixed the definition of the x1, x2, y1, y2 metadata (7cab798). These were specified as the coordinates of the lower-left and upper-right _pixels_, instead of the lower-left and upper-right _corners_ of the domain. 

From my side, this pull request is now ready to be merged into master.

Edit: the build for this branch is failing because of 4e63e93 (in fact, the build passes for the PR)

dnerini committed:
Fix test argument

dnerini committed:
Include quality field

dnerini committed:
Fix extent redefinition

RubenImhoff commented:
Nice work, Daniele, this looks good now! I'm glad we're finally able to add this.

RubenImhoff commented on pysteps/io/importers.py:
```diff
@@ -314,33 +343,17 @@ def import_saf_crri(filename, **kwargs):
+            data = np.array(ds_rainfall.variables["crr_intensity"][idx_y, idx_x])
+        else:
+            data = np.array(ds_rainfall.variables["crr_intensity"])        precipitation = np.where(data == 65535, np.nan, data)
+            precipitation = np.where(data == 65535, np.nan, data)
```
I've quickly scanned through most of the changes, and I'm not sure if this is correct, but it may also be a result of reading through it online.
Shouldn't it be:

```
if idx_x is not None:
    data = np.array(...)
else:
    data = np.array(...)

precipitation = np.where(...)
```

dnerini replied:
you're absolutely right, the error was fixed by a4352e8

RubenImhoff committed:
Update interface.py

To make it alphabetical again. ;)

dnerini committed:
Add online reference
